OCPBUGS-64793: Infrastructure Topology calculation is wrong

barbacbd opened the pull request:
If masters are schedulable then all nodes, not just non-controlplane nodes, should
    be taken into account when deciding what we tell OLM operators about the topology.
    
    Example:
    1.Deploy a cluster with 3 CP nodes (schedulable) and 1 Worker Node
    2.Inspect infrastructureTopology attribute of the cluster
    
    Expect: infrastructureTopology= HighlyAvailableTopologyMode

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Control plane node scheduling now correctly enables schedulability for deployments with fewer than two compute replicas.
  * Improved determination of infrastructure topology for clusters with very small worker counts.
  * Aligned topology handling so control plane and infrastructure topology decisions are consistent for small cluster configurations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

barbacbd committed:
OCPBUGS-64793: Infrastructure Topology calculation is wrong

If masters are schedulable then all nodes, not just non-controlplane nodes, should
be taken into account when deciding what we tell OLM operators about the topology.

Example:
1.Deploy a cluster with 3 CP nodes (schedulable) and 1 Worker Node
2.Inspect infrastructureTopology attribute of the cluster

Expect: infrastructureTopology= HighlyAvailableTopologyMode

barbacbd commented:
/jira refresh

tthvo commented on pkg/asset/manifests/topologies.go:
```diff
@@ -39,8 +39,6 @@ func determineTopologies(installConfig *types.InstallConfig) (controlPlaneTopolo
 			infrastructureTopology = controlPlaneTopology
 		}
-	case 1:
-		infrastructureTopology = configv1.SingleReplicaTopologyMode
```
:thought_balloon:  Is there a case where control plane nodes are non-schedulable in the 3 ctrlplane + 1 worker setup?

sadasu replied:
Cuurently when numOfWorker == 1, masters are not schedulable. See https://github.com/openshift/installer/blob/main/pkg/asset/manifests/scheduler.go#L75

So, even by setting `infrastructureTopology` to a different value, we will not have enough nodes to have a successful cluster.

tthvo replied:
oh right, then the bug is not valid? I actually have 0 clue now :sweat_smile:

gpei replied:
And we have some ongoing discussions about this PR in Slack https://redhat-internal.slack.com/archives/C01V1DP387R/p1768539319722709?thread_ts=1768472465.719819&cid=C01V1DP387R . Any input from you would be greatly appreciated.

sadasu commented on pkg/asset/manifests/ingress_test.go:
```diff
@@ -187,14 +187,14 @@ func TestGenerateIngerssDefaultPlacement(t *testing.T) {
 			installConfigBuildOptions: []icOption{icBuild.forNone()},
 			controlPlaneTopology:      configv1.HighlyAvailableTopologyMode,
-			infrastructureTopology:    configv1.SingleReplicaTopologyMode,
+			infrastructureTopology:    configv1.HighlyAvailableTopologyMode,
```
When there is a single control-plane node, we should end up with `controlPlaneTopology==configv1.SingleReplicaTopologyMode` and `infrastructureTopology==configv1.SingleReplicaTopologyMode`. I don't see a test for that.

tthvo commented:
/test e2e-aws-ovn-upi e2e-aws-ovn-single-node

tthvo commented on pkg/asset/manifests/topologies.go:
```diff
@@ -28,6 +28,11 @@ func determineTopologies(installConfig *types.InstallConfig) (controlPlaneTopolo
+		// Control planes are schedulable when there are < 2 workers.
+		// Adjust the number of schedulable nodes here to reflect.
+		numOfWorkers += controlPlaneReplicas
+	}
```
Also, what about `case 1: ` below?

https://github.com/openshift/installer/blob/4180662515e75d11d0a7984c66d574f3c64a765a/pkg/asset/manifests/topologies.go#L42-L43

It is setting infrastructure topology is `configv1.SingleReplicaTopologyMode `. This new check will also make the `case 1: ` unreachable unless SNO.

tthvo reviewed (commented):
Based on [comment](https://issues.redhat.com/browse/OCPBUGS-56954?focusedId=28466962&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-28466962), the change worked for me. The question I have above is really just a matter of cleanup (i.e. remove unreachable case if needed).

```bash
$ yq .controlPlane install-config.yaml 
architecture: amd64
hyperthreading: Enabled
name: master
platform: {}
replicas: 3

$ yq .compute[0] install-config.yaml 
architecture: amd64
hyperthreading: Enabled
name: worker
platform: {}
replicas: 1

$ ./openshift-install create manifests --dir=.
...output-omitted...

$ yq .spec.mastersSchedulable manifests/cluster-scheduler-02-config.yml 
true

$ yq .status.infrastructureTopology manifests/cluster-infrastructure-02-config.yml 
HighlyAvailable
```

/lgtm

zaneb commented on pkg/asset/manifests/topologies.go:
```diff
@@ -28,6 +28,11 @@ func determineTopologies(installConfig *types.InstallConfig) (controlPlaneTopolo
+	}
 
 	switch numOfWorkers {
 	case 0:
```
I think this change is effectively applying the same rules that we use for 2-node/arbiter case to everything. So it's valid, but the comments and code below that suggest there's still a special case that applies only to 2-node/arbiter clusters is now misleading. Better to delete it I think.

sadasu commented:
> In pre-merge testing, I noticed another change in behavior. When we specify a single control-plane node and 1 compute nodes, the control-plane will also be set to schedulable. Then this results in a cluster with `controlPlaneTopology: SingleReplica` and `infrastructureTopology: HighlyAvailable`. Is this expected?

Good point. I don't think we want that. As part of this fix, we want `infrastructureTopology: HighlyAvailable` to happen when `controlPlaneTopology` is also `HighlyAvailable`.

Afaik, https://github.com/openshift/installer/pull/10077#discussion_r2700010786 should take of that.

oarribas commented:
/remove-lifecycle stale

barbacbd commented:
I've addressed the review comments:

1. **Simplified the switch to if/else** (zaneb's comment): The topology determination logic is now a simple if/else since we can only have  (Single) or  (HA) after adding control plane replicas when needed.

2. **Removed unreachable code in case 1**: The nested if check for `controlPlaneTopology == HighlyAvailableTopologyMode` inside `case 1` was unreachable because when we have an HA control plane (3+ nodes) and < 2 workers, we add 3+ CP nodes to numOfWorkers, making it impossible to hit case 1.

3. **Test coverage for 1CP+0W**: The test at line 31-40 in topologies_test.go already covers this scenario and passes.

4. **Updated Jira target version to 5.0.0**: The issue was targeting 4.22.0 but this PR is against master (5.0.0). I've updated the Jira ticket accordingly.

The logic now correctly handles all scenarios:
- **1 CP + 0 W**: numOfWorkers=0 → adds 1 CP → numOfWorkers=1 → Single ✓ (SNO)
- **3 CP + 0 W**: numOfWorkers=0 → adds 3 CP → numOfWorkers=3 → HA ✓ (compact cluster)
- **3 CP + 1 W**: numOfWorkers=1 → adds 3 CP → numOfWorkers=4 → HA ✓ (bug fix - this was the issue)
- **1 CP + 1 W**: numOfWorkers=1 → adds 1 CP → numOfWorkers=2 → HA ✓ (masters schedulable)

All tests pass.

barbacbd commented:
/jira refresh
